### PR TITLE
Added documentation to Decoupled, Conditionals, Counter

### PR DESCRIPTION
**Pull request by azidar**
**Related issue**: #411

**Type of change**: other enhancement

**Impact**: no functional change, add deprecations

**Development Phase**: implementation

**Release Notes**

Open to feedback about this ScalaDoc.

### CODE REVIEW

File: src/main/scala/chisel3/util/Counter.scala [hunk 1]
```diff
@@ -59,6 +91,11 @@ object Counter
   @chiselName
   def apply(cond: Bool, n: Int): (UInt, Bool) = {
     val c = new Counter(n)
+    // Note this use of var is generally frowned upon! Don't use this as an example.
+    // This is done because we wanted the hardware generated by c.inc() to be wrapped
+    // in a when statement, but still needed to refer to the final value returned by
+    // c.inc() so we could return cond && wrap. Unless you really, really know what
+    // you are doing, IGNORE THIS CODE!!!!!
     var wrap: Bool = null
```
Comment (edwardcwang) [inline src/main/scala/chisel3/util/Counter.scala]: I'm curious as to how this code avoids a NullPointerException...

Comment (azidar) [reply src/main/scala/chisel3/util/Counter.scala]: It always assigns to `wrap` the return value of `c.inc()`

**Comment by ucbjrl** [comment]
Compile errors:
```
[info] Compiling 34 Scala sources to /vm/home/jenkins/sharedspace/chisel3_2/target/scala-2.12/classes ...
[error] /vm/home/jenkins/sharedspace/chisel3_2/src/main/scala/chisel3/util/Counter.scala:9:1: expected start of definition
[error] type Counter = Counter.Counter
[error] ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
Makefile:33: recipe for target 'smoke' failed
```

**Comment by edwardcwang** [comment]
retest this please

**Comment by ducky64** [comment]
retest this please

**Comment by chick** [comment]
retest this please

**Comment by ducky64** [comment]
Poke, what's going on with this PR? I think this is the last PR for #411?

**Comment by ucbjrl** [comment]
I've backed out the code changes to `util.Counter`. We can get this to successfully compile by moving the deprecation warning and the type definition into the util package object (see f61bddd), but then we fail tests using `Counter` with a `ScalaReflectionException`:
```
Caused by: scala.ScalaReflectionException: type Counter is not a class
	at scala.reflect.api.Symbols$SymbolApi.asClass(Symbols.scala:272)
	at scala.reflect.api.Symbols$SymbolApi.asClass$(Symbols.scala:272)
 ...
	at chisel3.Bundle.cloneType(Aggregate.scala:779)
	at chisel3.Bundle.cloneType(Aggregate.scala:670)
	at chisel3.Data.cloneTypeFull(Data.scala:472)
	at chisel3.experimental.IO$.apply(Module.scala:111)
	at chisel3.experimental.BaseModule.IO(Module.scala:348)
	at chiselTests.InvalidateAPISpec$ModuleWithIncompleteAssignment$1.<init>(InvalidateAPISpec.scala:121)
	at chiselTests.InvalidateAPISpec.$anonfun$new$25(InvalidateAPISpec.scala:130)
```
This appears to only happen in the `InvalidateAPISpec.scala` tests using `Counter`. Other tests with `Counter` are ok.

File: src/main/scala/chisel3/util/Counter.scala [hunk 2]
```diff
@@ -5,10 +5,12 @@ package chisel3.util
 import chisel3._
 import chisel3.internal.naming.chiselName  // can't use chisel3_ version because of compile order
 
-/** A counter module
+/** Used to generate an inline hardware counter
   *
   * Typically instantiated with apply methods in [[Counter$ object Counter]]
   *
+  * Does not create a new Chisel Module
```
Comment (ducky64) [inline src/main/scala/chisel3/util/Counter.scala]: You could merge this with the above sentence: 
`Used to generate an inline (logic directly in the containing Module, no internal Module is created) hardware counter.`

Comment (chick) [reply src/main/scala/chisel3/util/Counter.scala]: Done. Commit added to PR

**Pull request merged**